lib/std/sync/futex/generic.zig

```diff
@@ -0,0 +1,274 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2015-2021 Zig Contributors
+// This file is part of [zig](https://ziglang.org/), which is MIT licensed.
+// The MIT license requires this copyright notice to be included in all copies
+// and substantial portions of the software.
+
+const std = @import("../../std.zig");
+const atomic = @import("../atomic.zig");
+
+const builtin = std.builtin;
+const assert = std.debug.assert;
+const helgrind: ?type = if (builtin.valgrind_support) std.valgrind.helgrind else null;
+
+pub fn Futex(comptime Event: type) type {
+    return struct {
+        const WaitLock = Event.Lock;
+        const bucket_count = Event.bucket_count;
+
+        const WaitBucket = struct {
+            lock: WaitLock = .{},
+            waiters: usize = 0,
+            tree: WaitTree = .{},
+
+            var array = [_]WaitBucket{WaitBucket{}} ** bucket_count;
+
+            /// Hash a address to a wait-bucket.
+            /// This uses the same method as seen in Amanieu's port of WTF::ParkingLot:
+            /// https://github.com/Amanieu/parking_lot/blob/master/core/src/parking_lot.rs
+            fn from(address: usize) *WaitBucket {
+                const seed = @truncate(usize, 0x9E3779B97F4A7C15);
```

Comment:
huh... googling this number shows that it is `2^64 * phi` and seems to have a long history of use in wait queues

e.g. https://softwareengineering.stackexchange.com/a/402543